{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Reference sequence mapping horserace\n",
    "\n",
    "ipyrad is capable of incorporating a reference sequence to aid in the assembly. There are actually 4 different assembly methods, 3 of which use reference sequence in some way. Here we test reference assisted assembly for ipyrad, stacks, and ddocent. Though aftrRAD performs nicely on empirical data and does allow for reference assisted assembly, we consider runtimes to be prohibitive and so exclude it from analysis here.\n",
    "\n",
    "Ideas for datasets (all have data in SRA):\n",
    "\n",
    "    Selection and sex-biased dispersal in a coastal shark: the influence of philopatry on adaptive variation\n",
    "    - 134 individuals (paper assembled w/ ddocent)\n",
    "    \n",
    "    Genome-wide data reveal cryptic diversity and genetic introgression in an Oriental cynopterine fruit bat radiation\n",
    "    - < 45 samples, 2 reference genomes in the same family\n",
    "    \n",
    "    Beyond the Coral Triangle: high genetic diversity and near panmixia in Singapore's populations of the broadcast spawning sea star Protoreaster nodosus\n",
    "    - 77 samples, it's a passerine, so there must be something reasonably close\n",
    "    \n",
    "    PSMC (pairwise sequentially Markovian coalescent) analysis of RAD (restriction site associated DNA) sequencing data\n",
    "    - 17 sticklebacks, they used stacks\n",
    "    \n",
    "For this analysis we chose the paired-end ddRAD dataset from Lah et al 2016 (http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0162792#sec002):\n",
    "    Spatially Explicit Analysis of Genome-Wide SNPs Detects Subtle Population Structure in a Mobile Marine Mammal, the Harbor Porpoise\n",
    "    - 49 samples from 3 populations of European Harbor Porpoise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import ipyrad as ip\n",
    "import shutil\n",
    "import glob\n",
    "import sys\n",
    "import os\n",
    "\n",
    "## Set the default directories for exec and data. \n",
    "WORK_DIR=\"/home/iovercast/manuscript-analysis/\"\n",
    "REFMAP_EMPIRICAL_DIR=os.path.join(WORK_DIR, \"Phocoena_empirical/\")\n",
    "REFMAP_FASTQS=os.path.join(REFMAP_EMPIRICAL_DIR, \"Final_Files_forDryad/Bbif_ddRADseq/fastq/\")\n",
    "IPYRAD_DIR=os.path.join(WORK_DIR, \"ipyrad/\")\n",
    "STACKS_DIR=os.path.join(WORK_DIR, \"stacks/\")\n",
    "DDOCENT_DIR=os.path.join(WORK_DIR, \"dDocent/\")\n",
    "\n",
    "## (emprical data dir will be created for us when we untar it)\n",
    "for dir in [WORK_DIR, REFMAP_EMPIRICAL_DIR, IPYRAD_DIR, STACKS_DIR, DDOCENT_DIR]:\n",
    "    if not os.path.exists(dir):\n",
    "        os.makedirs(dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch the Phocoena raw sequence data \n",
    "We will use the sra-toolkit command `fastq-dump` to pull the PE reads out of SRA. \n",
    "This maybe isn't the best way, or the quickest, but it'll get the job done. Takes \n",
    "~30 minutes and requires ~70GB of space. After I downloaded the fq I looked at a\n",
    "couple random samples in FastQC to get an idea where to trim in step2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing 662\tDoing 663\tDoing 664\tDoing 665\tDoing 666\tDoing 667\tDoing 668\tDoing 669\tDoing 670\tDoing 671\tDoing 672\tDoing 673\tDoing 674\tDoing 675\tDoing 676\tDoing 677\tDoing 678\tDoing 679\tDoing 680\tDoing 681\tDoing 682\tDoing 683\tDoing 684\tDoing 685\tDoing 686\tDoing 687\tDoing 688\tDoing 689\tDoing 690\tDoing 691\tDoing 692\tDoing 693\tDoing 694\tDoing 695\tDoing 696\tDoing 697\tDoing 698\tDoing 699\tDoing 700\tDoing 701\tDoing 702\t"
     ]
    }
   ],
   "source": [
    "os.chdir(REFMAP_EMPIRICAL_DIR)\n",
    "!mkdir raws\n",
    "!cd raws\n",
    "## Grab the sra-toolkit pre-built binaries to download from SRA\n",
    "## This works, but commented for now so it doesn't keep redownloading\n",
    "!wget http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.8.0/sratoolkit.2.8.0-ubuntu64.tar.gz\n",
    "!tar -xvzf sratoolkit*\n",
    "FQ_DUMP = os.path.join(REFMAP_EMPIRICAL_DIR, \"sratoolkit.2.8.0-ubuntu64/bin/fastq-dump\")\n",
    "res = subprocess.check_output(FQ_DUMP + \" -version\", shell=True)\n",
    "\n",
    "## The SRR numbers for the samples from this bioproject range from SRR4291662 to SRR4291705\n",
    "## so go fetch them one by one\n",
    "for samp in range(662, 706):\n",
    "    print(\"Doing {}\\t\".format(samp)),\n",
    "    res = subprocess.check_output(FQ_DUMP + \" --split-files SRR4291\" + str(samp), shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## The SRA download files have wonky names, like SRR1234_R1.fastq.gz, but ipyrad expects SRR1234_R1_.fastq.gz,\n",
    "## so we have to fix the filenames. Filename hax...\n",
    "import glob\n",
    "for f in glob.glob(REFMAP_EMPIRICAL_DIR + \"raws/*.fastq.gz\"):\n",
    "    splits = f.split(\"/\")[-1].split(\"_\")\n",
    "    newf = REFMAP_EMPIRICAL_DIR + \"raws/\" + splits[0] + \"_R\" + splits[1].split(\".\")[0] + \"_.fastq.gz\"\n",
    "    os.rename(f, newf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch the bottlenose dolphin genome\n",
    "Tursiops truncatus reference genome. Divergence time between dolphin and porpoise is approximately 15Mya, which is on the order of divergence between humans and orang. There is also a genome for the Minke whale, which is much more deeply diverged (~30Mya), could be interesting to try both to see how it works.\n",
    "\n",
    "Minke whale - http://www.nature.com/ng/journal/v46/n1/full/ng.2835.html#accessions\n",
    "\n",
    "SRA Data table for converting fq files to sample name as used in the paper: file:///home/chronos/u-b20882bda0f801c7265d4462f127d0cb4376d46d/Downloads/Tune/SraRunTable.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2016-12-18 10:29:57--  ftp://ftp.ensembl.org/pub/release-87/fasta/tursiops_truncatus/dna/Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz\n",
      "           => ‘Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz’\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.203.85\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.203.85|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-87/fasta/tursiops_truncatus/dna ... done.\n",
      "==> SIZE Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz ... 453168562\n",
      "==> PASV ... done.    ==> RETR Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz ... done.\n",
      "Length: 453168562 (432M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 453,168,562 2.88MB/s   in 3m 33s \n",
      "\n",
      "2016-12-18 10:33:32 (2.03 MB/s) - ‘Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz’ saved [453168562]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.chdir(REFMAP_EMPIRICAL_DIR)\n",
    "!mkdir TurtrunRef\n",
    "!cd TurtrunRef\n",
    "!wget ftp://ftp.ensembl.org/pub/release-87/fasta/tursiops_truncatus/dna/Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz\n",
    "\n",
    "## Ensembl distributes gzip'd reference sequence files, but samtools really wants it to be bgzipped or uncompressed\n",
    "!gunzip Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trim reads w/ cutadapt\n",
    "To reduce any potential bias introduced by differences in trimming and filtering methods we will trim and filter the raw reads w/ cutadapt, and use this QC'd dataset as the starting point for assembly for all 3 programs. If you are wondering, you __can__ run fastqc from the command line, like this\n",
    "\n",
    "    fastqc -o fastqc_out/ SRR4291662_1.fastq SRR4291662_2.fastq\n",
    "    \n",
    "We'll trim R1 and R2 to 85bp following Lah et al 2016. The `-l` flag for cutadapt specifies the length to which each read will be trimmed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR4291662_R1_.fastq.gz\n",
      "SRR4291662_R2_.fastq.gz\n",
      "SRR4291663_R1_.fastq.gz\n",
      "SRR4291663_R2_.fastq.gz\n",
      "SRR4291664_R1_.fastq.gz\n",
      "SRR4291664_R2_.fastq.gz\n",
      "SRR4291665_R1_.fastq.gz\n",
      "SRR4291665_R2_.fastq.gz\n",
      "SRR4291666_R1_.fastq.gz\n",
      "SRR4291666_R2_.fastq.gz\n",
      "SRR4291667_R1_.fastq.gz\n",
      "SRR4291667_R2_.fastq.gz\n",
      "SRR4291668_R1_.fastq.gz\n",
      "SRR4291668_R2_.fastq.gz\n",
      "SRR4291669_R1_.fastq.gz\n",
      "SRR4291669_R2_.fastq.gz\n",
      "SRR4291670_R1_.fastq.gz\n",
      "SRR4291670_R2_.fastq.gz\n",
      "SRR4291671_R1_.fastq.gz\n",
      "SRR4291671_R2_.fastq.gz\n",
      "SRR4291672_R1_.fastq.gz\n",
      "SRR4291672_R2_.fastq.gz\n",
      "SRR4291673_R1_.fastq.gz\n",
      "SRR4291673_R2_.fastq.gz\n",
      "SRR4291674_R1_.fastq.gz\n",
      "SRR4291674_R2_.fastq.gz\n",
      "SRR4291675_R1_.fastq.gz\n",
      "SRR4291675_R2_.fastq.gz\n",
      "SRR4291676_R1_.fastq.gz\n",
      "SRR4291676_R2_.fastq.gz\n",
      "SRR4291677_R1_.fastq.gz\n",
      "SRR4291677_R2_.fastq.gz\n",
      "SRR4291678_R1_.fastq.gz\n",
      "SRR4291678_R2_.fastq.gz\n",
      "SRR4291679_R1_.fastq.gz\n",
      "SRR4291679_R2_.fastq.gz\n",
      "SRR4291680_R1_.fastq.gz\n",
      "SRR4291680_R2_.fastq.gz\n",
      "SRR4291681_R1_.fastq.gz\n",
      "SRR4291681_R2_.fastq.gz\n",
      "SRR4291682_R1_.fastq.gz\n",
      "SRR4291682_R2_.fastq.gz\n",
      "SRR4291683_R1_.fastq.gz\n",
      "SRR4291683_R2_.fastq.gz\n",
      "SRR4291684_R1_.fastq.gz\n",
      "SRR4291684_R2_.fastq.gz\n",
      "SRR4291685_R1_.fastq.gz\n",
      "SRR4291685_R2_.fastq.gz\n",
      "SRR4291686_R1_.fastq.gz\n",
      "SRR4291686_R2_.fastq.gz\n",
      "SRR4291687_R1_.fastq.gz\n",
      "SRR4291687_R2_.fastq.gz\n",
      "SRR4291688_R1_.fastq.gz\n",
      "SRR4291688_R2_.fastq.gz\n",
      "SRR4291689_R1_.fastq.gz\n",
      "SRR4291689_R2_.fastq.gz\n",
      "SRR4291690_R1_.fastq.gz\n",
      "SRR4291690_R2_.fastq.gz\n",
      "SRR4291691_R1_.fastq.gz\n",
      "SRR4291691_R2_.fastq.gz\n",
      "SRR4291692_R1_.fastq.gz\n",
      "SRR4291692_R2_.fastq.gz\n",
      "SRR4291693_R1_.fastq.gz\n",
      "SRR4291693_R2_.fastq.gz\n",
      "SRR4291694_R1_.fastq.gz\n",
      "SRR4291694_R2_.fastq.gz\n",
      "SRR4291695_R1_.fastq.gz\n",
      "SRR4291695_R2_.fastq.gz\n",
      "SRR4291696_R1_.fastq.gz\n",
      "SRR4291696_R2_.fastq.gz\n",
      "SRR4291697_R1_.fastq.gz\n",
      "SRR4291697_R2_.fastq.gz\n",
      "SRR4291698_R1_.fastq.gz\n",
      "SRR4291698_R2_.fastq.gz\n",
      "SRR4291699_R1_.fastq.gz\n",
      "SRR4291699_R2_.fastq.gz\n",
      "SRR4291700_R1_.fastq.gz\n",
      "SRR4291700_R2_.fastq.gz\n",
      "SRR4291701_R1_.fastq.gz\n",
      "SRR4291701_R2_.fastq.gz\n",
      "SRR4291702_R1_.fastq.gz\n",
      "SRR4291702_R2_.fastq.gz\n",
      "SRR4291703_R1_.fastq.gz\n",
      "SRR4291703_R2_.fastq.gz\n",
      "SRR4291704_R1_.fastq.gz\n",
      "SRR4291704_R2_.fastq.gz\n",
      "SRR4291705_R1_.fastq.gz\n",
      "SRR4291705_R2_.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘trimmed’: File exists\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n",
      "cutadapt version 1.11\n",
      "Copyright (C) 2010-2016 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. The reverse complement is *not* automatically\n",
      "searched. All reads from input.fastq will be written to output.fastq with the\n",
      "adapter sequence removed. Adapter matching is error-tolerant. Multiple adapter\n",
      "sequences can be given (use further -a options), but only the best-matching\n",
      "adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Use \"cutadapt --help\" to see all command-line options.\n",
      "See http://cutadapt.readthedocs.org/ for full documentation.\n",
      "\n",
      "cutadapt: error: no such option: -l\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$REFMAP_EMPIRICAL_DIR\"\n",
    "cd $1\n",
    "mkdir trimmed\n",
    "for i in `ls raws`; do echo $i; cutadapt -l 10 raws/$i > trimmed/$i; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  New Assembly: refmap-empirical\n",
      "ipyrad -p params-refmap-empirical.txt -s 1 --force\n",
      "\n",
      " -------------------------------------------------------------\n",
      "  ipyrad [v.0.5.10]\n",
      "  Interactive assembly and analysis of RAD-seq data\n",
      " -------------------------------------------------------------\n",
      "  New Assembly: refmap-empirical\n",
      "  local compute node: [40 cores] on node001\n",
      "\n",
      "  Step 1: Loading sorted fastq data to Samples\n",
      "\n",
      "  Encountered an unexpected error (see ./ipyrad_log.txt)\n",
      "  Error message is below -------------------------------\n",
      "float division by zero\n",
      "\n",
      "\n",
      "real\t0m12.615s\n",
      "user\t0m1.395s\n",
      "sys\t0m0.344s\n"
     ]
    }
   ],
   "source": [
    "IPYRAD_REFMAP_DIR = os.path.join(REFMAP_EMPIRICAL_DIR, \"ipyrad\")\n",
    "if not os.path.exists(IPYRAD_REFMAP_DIR):\n",
    "    os.makedirs(IPYRAD_REFMAP_DIR)\n",
    "os.chdir(IPYRAD_REFMAP_DIR)\n",
    "\n",
    "## Make a new assembly and set some assembly parameters\n",
    "data = ip.Assembly(\"refmap-empirical\")\n",
    "data.set_params(\"sorted_fastq_path\", REFMAP_EMPIRICAL_DIR + \"raws/*.fastq.gz\")\n",
    "data.set_params(\"project_dir\", \"reference-assembly\")\n",
    "data.set_params(\"assembly_method\", \"reference\")\n",
    "data.set_params(\"reference_sequence\", REFMAP_EMPIRICAL_DIR + \"TurtrunRef/Tursiops_truncatus.turTru1.dna_rm.toplevel.fa.gz\")\n",
    "data.set_params(\"datatype\", \"pairddrad\")\n",
    "data.set_params(\"restriction_overhang\", (\"TGCAG\", \"CGG\"))\n",
    "data.set_params(\"clust_threshold\", 0.9)\n",
    "data.set_params('max_low_qual_bases', 30)\n",
    "data.set_params('min_samples_locus', 3)\n",
    "data.set_params('trim_overhang', (0,15,0,15))\n",
    "\n",
    "data.write_params(force=True)\n",
    "\n",
    "cmd = \"ipyrad -p params-refmap-empirical.txt -s 1 --force\".format(dir)\n",
    "print(cmd)\n",
    "!time $cmd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do ipyrad refmap empirical"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do Stacks refmap empirical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "STACKS_REFMAP_DIR = os.path.join(REFMAP_EMPIRICAL_DIR, \"stacks\")\n",
    "if not os.path.exists(STACKS_REFMAP_DIR):\n",
    "    os.makedirs(STACKS_REFMAP_DIR)\n",
    "os.chidir(STACKS_REFMAP_DIR)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Do dDocent refmap empirical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DDOCENT_REFMAP_DIR = os.path.join(REFMAP_EMPIRICAL_DIR, \"ddocent\")\n",
    "if not os.path.exists(DDOCENT_REFMAP_DIR):\n",
    "    os.makedirs(DDOCENT_REFMAP_DIR)\n",
    "os.chidir(DDOCENT_REFMAP_DIR)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
